src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs

```diff
@@ -0,0 +1,159 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Diagnostics;
+using Microsoft.AspNetCore.Mvc.Razor.Compilation;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Internal
+{
+    internal static class CompilationFailedResultFactory
```

Comment:
There's a bit of overlap in terms of creating CompilationFailedResult from both Razor and Roslyn. Combining the two here.